File: lib/services/rsa_service.dart
import 'dart:convert';
import 'dart:typed_data';
import 'package:pointycastle/export.dart';

class RsaService {
  static const String _publicKeyPem = '''
-----BEGIN PUBLIC KEY-----
MIIBIjANBgkqhkiG9w0BAQEFAAOCAQ8AMIIBCgKCAQEA4s6sEpFv6IIu4goE2Qc6
euzyElJsvAUVJVfnXRljjsSq8qiFVHJpGUlFzI4IWOAz8GTk65i3fxLmbuKos2QD
+pxql2TluJ4lbvZXRqotAwVqilnDk52vW8OP6un/u+gG+2Wyhjsn67YNCBrXsJ8C
O+d7zcR7vcjpXT6oXyWqw3ePwVXTLOWonKpGvTS/ypzY0cDPOHv/x80lXguQj65U
qPy9eVljeSo8ackIc0ylM44weqIENTWFnUW05ueW1zrtUAwOCVyRjCPW5iNafy7B
Venhh8Htxfn6NT57fTPgDTMqMD6BB3U0PgpYjm4mFIPuP7YMUbfD6L3U4QY/COjL
EQIDAQAB
-----END PUBLIC KEY-----''';

  static RSAPublicKey _parsePublicKey() {
    final lines = _publicKeyPem
        .split('\n')
        .where((line) =>
            !line.startsWith('-----BEGIN') && !line.startsWith('-----END'))
        .join();
    final keyBytes = base64Decode(lines);

    // PKCS#8 SubjectPublicKeyInfo structure:
    //   SEQUENCE {
    //     SEQUENCE { algorithm OID, NULL }
    //     BIT STRING {
    //       SEQUENCE {
    //         INTEGER (modulus)
    //         INTEGER (exponent)
    //       }
    //     }
    //   }
    final outer = _derReadSequence(keyBytes, 0);
    final algoSeqEnd = _derSkipObject(keyBytes, outer.contentOffset);
    final bitString = _derReadTag(keyBytes, algoSeqEnd);

    // BIT STRING has a leading 0x00 "unused bits" byte
    final rsaKeyOffset = bitString.contentOffset + 1;
    final inner = _derReadSequence(keyBytes, rsaKeyOffset);
    final modulus = _derReadInteger(keyBytes, inner.contentOffset);
    final exponent = _derReadInteger(keyBytes, modulus.nextOffset);

    return RSAPublicKey(modulus.value, exponent.value);
  }

  static String encrypt(String plainText) {
    final publicKey = _parsePublicKey();
    final encryptor = OAEPEncoding(RSAEngine())
      ..init(true, PublicKeyParameter<RSAPublicKey>(publicKey));
    final plainBytes = Uint8List.fromList(utf8.encode(plainText));
    final encrypted = encryptor.process(plainBytes);
    return base64Encode(encrypted);
  }

  // --- Minimal DER parser helpers ---

  static ({int contentOffset, int contentLength, int nextOffset}) _derReadTag(
      Uint8List data, int offset) {
    // skip tag byte
    int pos = offset + 1;
    int length = data[pos++];
    if (length & 0x80 != 0) {
      final numBytes = length & 0x7F;
      length = 0;
      for (int i = 0; i < numBytes; i++) {
        length = (length << 8) | data[pos++];
      }
    }
    return (
      contentOffset: pos,
      contentLength: length,
      nextOffset: pos + length,
    );
  }

  static ({int contentOffset, int contentLength}) _derReadSequence(
      Uint8List data, int offset) {
    assert(data[offset] == 0x30); // SEQUENCE tag
    final tag = _derReadTag(data, offset);
    return (contentOffset: tag.contentOffset, contentLength: tag.contentLength);
  }

  static int _derSkipObject(Uint8List data, int offset) {
    final tag = _derReadTag(data, offset);
    return tag.nextOffset;
  }

  static ({BigInt value, int nextOffset}) _derReadInteger(
      Uint8List data, int offset) {
    assert(data[offset] == 0x02); // INTEGER tag
    final tag = _derReadTag(data, offset);
    final bytes =
        data.sublist(tag.contentOffset, tag.contentOffset + tag.contentLength);
    // Parse as unsigned big-endian
    BigInt value = BigInt.zero;
    for (final b in bytes) {
      value = (value << 8) | BigInt.from(b);
    }
    return (value: value, nextOffset: tag.nextOffset);
  }
}

File: ios-swift/eNagarsewa/Services/Security/RsaService.swift
import Foundation
import Security

/// RSA-OAEP(SHA-1) encryption using the server's public key — matches Flutter's RsaService
/// (pointycastle `OAEPEncoding(RSAEngine())`, which defaults to SHA-1).
/// Used only to encrypt the password / confirm-password fields before citizen self-registration.
enum RsaService {

    private static let publicKeyPEM = """
    -----BEGIN PUBLIC KEY-----
    MIIBIjANBgkqhkiG9w0BAQEFAAOCAQ8AMIIBCgKCAQEA4s6sEpFv6IIu4goE2Qc6
    euzyElJsvAUVJVfnXRljjsSq8qiFVHJpGUlFzI4IWOAz8GTk65i3fxLmbuKos2QD
    +pxql2TluJ4lbvZXRqotAwVqilnDk52vW8OP6un/u+gG+2Wyhjsn67YNCBrXsJ8C
    O+d7zcR7vcjpXT6oXyWqw3ePwVXTLOWonKpGvTS/ypzY0cDPOHv/x80lXguQj65U
    qPy9eVljeSo8ackIc0ylM44weqIENTWFnUW05ueW1zrtUAwOCVyRjCPW5iNafy7B
    Venhh8Htxfn6NT57fTPgDTMqMD6BB3U0PgpYjm4mFIPuP7YMUbfD6L3U4QY/COjL
    EQIDAQAB
    -----END PUBLIC KEY-----
    """

    enum RsaError: Error {
        case invalidKeyFormat
        case keyCreationFailed(String)
        case encryptionFailed(String)
    }

    static func encrypt(_ plainText: String) throws -> String {
        let pkcs1Data = try pkcs1PublicKeyData(fromSPKIPem: publicKeyPEM)

        let attributes: [CFString: Any] = [
            kSecAttrKeyType: kSecAttrKeyTypeRSA,
            kSecAttrKeyClass: kSecAttrKeyClassPublic,
        ]
        var cfError: Unmanaged<CFError>?
        guard let secKey = SecKeyCreateWithData(pkcs1Data as CFData, attributes as CFDictionary, &cfError) else {
            let message = cfError.map { (CFErrorCopyDescription($0.takeRetainedValue()) as String) } ?? "unknown"
            throw RsaError.keyCreationFailed(message)
        }

        let plainData = Data(plainText.utf8)
        guard SecKeyIsAlgorithmSupported(secKey, .encrypt, .rsaEncryptionOAEPSHA1) else {
            throw RsaError.encryptionFailed("OAEP-SHA1 not supported for this key")
        }

        var encError: Unmanaged<CFError>?
        guard let cipherData = SecKeyCreateEncryptedData(
            secKey, .rsaEncryptionOAEPSHA1, plainData as CFData, &encError
        ) as Data? else {
            let message = encError.map { (CFErrorCopyDescription($0.takeRetainedValue()) as String) } ?? "unknown"
            throw RsaError.encryptionFailed(message)
        }

        return cipherData.base64EncodedString()
    }

    // MARK: - Minimal DER parsing: extract the inner PKCS#1 RSAPublicKey TLV from an SPKI blob.
    //
    // SubjectPublicKeyInfo ::= SEQUENCE {
    //   SEQUENCE { algorithm OID, NULL }                 -- skipped
    //   BIT STRING {
    //     SEQUENCE { INTEGER modulus, INTEGER exponent } -- this is the PKCS#1 blob we need
    //   }
    // }

    private static func pkcs1PublicKeyData(fromSPKIPem pem: String) throws -> Data {
        let base64 = pem
            .split(separator: "\n")
            .filter { !$0.hasPrefix("-----BEGIN") && !$0.hasPrefix("-----END") }
            .joined()
        guard let keyBytes = Data(base64Encoded: base64) else { throw RsaError.invalidKeyFormat }
        let bytes = [UInt8](keyBytes)

        let outer = try derReadTag(bytes, offset: 0)                      // outer SEQUENCE
        let algoEnd = try derSkipObject(bytes, offset: outer.contentOffset) // skip algorithm SEQUENCE
        let bitString = try derReadTag(bytes, offset: algoEnd)             // BIT STRING

        // BIT STRING content starts with a leading 0x00 "unused bits" byte.
        let innerOffset = bitString.contentOffset + 1
        let inner = try derReadTag(bytes, offset: innerOffset)             // inner SEQUENCE (PKCS#1)

        let tlvStart = innerOffset
        let tlvEnd = inner.nextOffset
        guard tlvEnd <= bytes.count else { throw RsaError.invalidKeyFormat }
        return Data(bytes[tlvStart..<tlvEnd])
    }

    private struct DERTag {
        let contentOffset: Int
        let contentLength: Int
        let nextOffset: Int
    }

    private static func derReadTag(_ data: [UInt8], offset: Int) throws -> DERTag {
        guard offset + 1 < data.count else { throw RsaError.invalidKeyFormat }
        var pos = offset + 1
        var length = Int(data[pos]); pos += 1
        if length & 0x80 != 0 {
            let numBytes = length & 0x7F
            length = 0
            guard pos + numBytes <= data.count else { throw RsaError.invalidKeyFormat }
            for _ in 0..<numBytes {
                length = (length << 8) | Int(data[pos]); pos += 1
            }
        }
        let next = pos + length
        guard next <= data.count else { throw RsaError.invalidKeyFormat }
        return DERTag(contentOffset: pos, contentLength: length, nextOffset: next)
    }

    private static func derSkipObject(_ data: [UInt8], offset: Int) throws -> Int {
        try derReadTag(data, offset: offset).nextOffset
    }
}
